doc/debuginfod-find.1
'\"! tbl | nroff \-man
'\" t macro stdmacro

.de SAMPLE
.br
.RS 0
.nf
.nh
..
.de ESAMPLE
.hy
.fi
.RE
..

.TH DEBUGINFOD-FIND 1
.SH NAME
debuginfod-find \- request debuginfo-related data

.SH SYNOPSIS
.B debuginfod-find [\fIOPTION\fP]... debuginfo \fIBUILDID\fP

.B debuginfod-find [\fIOPTION\fP]... executable \fIBUILDID\fP

.B debuginfod-find [\fIOPTION\fP]... source \fIBUILDID\fP \fI/FILENAME\fP

.SH DESCRIPTION
\fBdebuginfod-find\fP queries one or more \fBdebuginfod\fP servers for
debuginfo-related data.  In case of a match, it saves the the
requested file into a local cache, prints the file name to standard
output, and exits with a success status of 0.  In case of any error,
it exits with a failure status and an error message to standard error.

.\" Much of the following text is duplicated with debuginfod.8

The debuginfod system uses buildids to identify debuginfo-related data.
These are stored as binary notes in ELF/DWARF files, and are
represented as lowercase hexadecimal.  For example, for a program
/bin/ls, look at the ELF note GNU_BUILD_ID:

.SAMPLE
% readelf -n /bin/ls | grep -A4 build.id
Note section [ 4] '.note.gnu.buildid' of 36 bytes at offset 0x340:
Owner          Data size  Type
GNU                   20  GNU_BUILD_ID
Build ID: 8713b9c3fb8a720137a4a08b325905c7aaf8429d
.ESAMPLE

Then the hexadecimal BUILDID is simply:

.SAMPLE
8713b9c3fb8a720137a4a08b325905c7aaf8429d
.ESAMPLE

In place of the hexadecimal buildid, debuginfod-find also accepts a path
name to to an ELF binary, from which it will then extract the buildid
and proceed.  In this case, pass the file name in place of the BUILDID
parameter, and ensure it has some character other than \fB[0-9a-f]\fP.
Files named \fBdeadbeef\fP can be passed with a \fB./deadbeef\fP extra
path component.

.SS debuginfo \fIBUILDID\fP

If the given buildid is known to a server, this request will result
in a binary object that contains the customary \fB.*debug_*\fP
sections.  This may be a split debuginfo file as created by
\fBstrip\fP, or it may be an original unstripped executable.

.SS executable \fIBUILDID\fP

If the given buildid is known to the server, this request will result
in a binary object that contains the normal executable segments.  This
may be a executable stripped by \fBstrip\fP, or it may be an original
unstripped executable.  \fBET_DYN\fP shared libraries are considered
to be a type of executable.

.SS source \fIBUILDID\fP \fI/SOURCE/FILE\fP

If the given buildid is known to the server, this request will result
in a binary object that contains the source file mentioned.  The path
should be absolute.  Relative path names commonly appear in the DWARF
file's source directory, but these paths are relative to
individual compilation unit AT_comp_dir paths, and yet an executable
is made up of multiple CUs.  Therefore, to disambiguate, debuginfod
expects source queries to prefix relative path names with the CU
compilation-directory, followed by a mandatory "/".

Note: the user should not elide \fB../\fP or \fB/./\fP or extraneous
\fB///\fP sorts of path components in the directory names, because if
this is how those names appear in the DWARF files, that is what
debuginfod needs to see too.

For example:
.TS
l l.
#include <stdio.h>	source BUILDID /usr/include/stdio.h
/path/to/foo.c	source BUILDID /path/to/foo.c
\../bar/foo.c AT_comp_dir=/zoo/	source BUILDID /zoo//../bar/foo.c
.TE

.SH "OPTIONS"

.TP
.B "\-v"
Increase verbosity, including printing frequent download-progress messages.


.SH "SECURITY"

debuginfod-find \fBdoes not\fP include any particular security
features.  It trusts that the binaries returned by the debuginfod(s)
are accurate.  Therefore, the list of servers should include only
trustworthy ones.  If accessed across HTTP rather than HTTPS, the
network should be trustworthy.  Authentication information through
the internal \fIlibcurl\fP library is not currently enabled, except
for the basic plaintext \%\fIhttp[s]://userid:password@hostname/\fP style.
(The debuginfod server does not perform authentication, but a front-end
proxy server could.)

.SH "ENVIRONMENT VARIABLES"

.TP 21
.B DEBUGINFOD_URLS
This environment variable contains a list of URL prefixes for trusted
debuginfod instances.  Alternate URL prefixes are separated by space.

.TP 21
.B DEBUGINFOD_TIMEOUT
This environment variable governs the timeout for each debuginfod HTTP
connection.  A server that fails to provide at least 100K of data
within this many seconds is skipped. The default is 90 seconds.  (Zero
or negative means "no timeout".)

.TP 21
.B DEBUGINFOD_CACHE_PATH
This environment variable governs the location of the cache where
downloaded files are kept.  It is cleaned periodically as this program
is reexecuted.  Cache management parameters may be set by files under
this directory: see the \fBdebuginfod_find_debuginfo(3)\fP man page
for details.  The default is $HOME/.debuginfod_client_cache.

.SH "FILES"
.LP
.PD .1v
.TP 20
.B $HOME/.debuginfod_client_cache
Default cache directory.
.PD

.SH "SEE ALSO"
.I "debuginfod(8)"
.I "debuginfod_find_debuginfod(3)"


doc/ChangeLog
2020-03-25  Frank Ch. Eigler  <fche@redhat.com>

	* debuginfod-find.1, debuginfod_find_debuginfo.3: Document
	acceptance of /path/names in place of buildids.

2020-03-22  Frank Ch. Eigler  <fche@redhat.com>

	* debuginfod_get_url.3: New function, documented ...
	* debuginfod_find_debuginfo.3: ... here.
	* Makefile.am (notrans_dist_*_man3): Add it.

2020-03-22  Frank Ch. Eigler  <fche@redhat.com>

	* debuginfod_get_user.3, _set_user.3: New functions, documented ...
	* debuginfod_find_debuginfo.3: ... here.
	* Makefile.am (notrans_dist_*_man3): List all debuginfod .3 functions.

2020-02-25  Frank Ch. Eigler  <fche@redhat.com>

	* debuginfod.8: Note that -R works just like -Z.rpm .

2020-02-25  Frank Ch. Eigler  <fche@redhat.com>

	* debuginfod.8: Document new --fdcache-prefetch option.

2020-02-05  Frank Ch. Eigler  <fche@redhat.com>

	* debuginfod.8: Document new -Z flag and tweak other bits.

2020-01-10  Mark Wielaard  <mark@klomp.org>

	* debuginfod_find_debuginfo.3 (DEBUGINFOD_PROGRESS): Mention progress
	output goes to stderr.

2020-01-11  Frank Ch. Eigler  <fche@redhat.com

	* debuginfod.8: Document --fdcache-fds and --fdcache-mbs opts.

2020-01-11  Frank Ch. Eigler  <fche@redhat.com>

	* debuginfod.8: Rework sections dealing with traversal/scanning,
	explaining new threading model.

2020-01-02  Mark Wielaard  <mark@klomp.org>

	* debuginfod.8 (DEBUGINFOD_TIMEOUT): Document as seconds to
	provide 100K, default 90.
	* debuginfod-find.1 (DEBUGINFOD_TIMEOUT): Likewise.
	* debuginfod_find_debuginfo.3 (DEBUGINFOD_TIMEOUT): Likewise.

2019-12-22  Frank Ch. Eigler  <fche@redhat.com

	* debuginfod.8: Add -U (DEB) flag, generalize RPM to "archive".

2019-12-04  Frank Ch. Eigler  <fche@redhat.com>

	* debuginfod-find.1: Bump default timeout to 30.
	* debuginfod_find_debuginfo.3: Ditto.
	Document $DEBUGINFOD_PROGRESS.

2019-11-26  Frank Ch. Eigler  <fche@redhat.com>
	    Aaron Merey  <amerey@redhat.com>

	* debuginfod.8, find-debuginfo.1, debuginfod_*.3: New files.

2019-09-02  Mark Wielaard  <mark@klomp.org>

	* readelf.1 (symbols): Add optional section name.
	(dyn-sym): Document new option.

2019-08-28  Mark Wielaard  <mark@klomp.org>

	* COPYING: Rename to...
	* COPYING-GFDL: ... this.

2019-08-23  Ben Woodard  <woodard@redhat.com>

	* Updated the eu-readelf man page to make it match the options
	that eu-readelf actually supports.

2019-08-22  Ben Woodard  <woodard@redhat.com>S

	* Move the .1 man pages to the correct place.

2019-08-21  Ben Woodard  <woodard@redhat.com>

	* Updated Changelog
	* Added README

2019-08-20  Ben Woodard  <woodard@redhat.com>

	* Added eu-elfclassify.1 man page based upon --help
	* Forked binutils readelf page to make eu-readelf.1 man page
	* Modified eu-readelf.1 to add -n:: option.
	* Disabled sgml file building per mjw.
	* Added man pages to Makefile.am

2019-06-20  Ben Woodard  <woodard@redhat.com>

	* Added the beginnings of some man pages

2019-08-21  Ben Woodard  <woodard@redhat.com>

	* Updated Changelog
	* Added README

2019-08-20  Ben Woodard  <woodard@redhat.com>

	* Added eu-elfclassify.1 man page based upon --help
	* Forked binutils readelf page to make eu-readelf.1 man page
	* Modified eu-readelf.1 to add -n:: option.
	* Disabled sgml file building per mjw.
	* Added man pages to Makefile.am

2019-06-20 Ben Woodard <woodard@redhat.com>

	* Added the beginnings of some man pages

2005-04-29  Ulrich Drepper  <drepper@redhat.com>

	* elfutils.sgml: Some typo fixes and a few extensions.
	Patch by Eric Christopher <echristo@redhat.com>.

2005-02-22  Ulrich Drepper  <drepper@redhat.com>

	* Makefile.am: Prefer pdf.

2003-08-11  Ulrich Drepper  <drepper@redhat.com>

	* Moved to CVS archive.


debuginfod/debuginfod-find.c
/* Command-line frontend for retrieving ELF / DWARF / source files
   from the debuginfod.
   Copyright (C) 2019-2020 Red Hat, Inc.
   This file is part of elfutils.

   This file is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 3 of the License, or
   (at your option) any later version.

   elfutils is distributed in the hope that it will be useful, but
   WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   General Public License for more details.

   You should have received copies of the GNU General Public License and
   the GNU Lesser General Public License along with this program.  If
   not, see <http://www.gnu.org/licenses/>.  */

#include "config.h"
#include "printversion.h"
#include "debuginfod.h"
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <argp.h>

#include <gelf.h>

/* Name and version of program.  */
ARGP_PROGRAM_VERSION_HOOK_DEF = print_version;

/* Bug report address.  */
ARGP_PROGRAM_BUG_ADDRESS_DEF = PACKAGE_BUGREPORT;

/* Short description of program.  */
static const char doc[] = N_("Request debuginfo-related content "
                             "from debuginfods listed in $" DEBUGINFOD_URLS_ENV_VAR ".");

/* Strings for arguments in help texts.  */
static const char args_doc[] = N_("debuginfo BUILDID\n"
                                  "executable BUILDID\n"
                                  "source BUILDID /FILENAME");

/* Definitions of arguments for argp functions.  */
static const struct argp_option options[] =
  {
   { "verbose", 'v', NULL, 0, "Increase verbosity.", 0 },
   { NULL, 0, NULL, 0, NULL, 0 }
  };

/* debuginfod connection handle.  */
static debuginfod_client *client;
static int verbose;

int progressfn(debuginfod_client *c __attribute__((__unused__)),
	       long a, long b)
{
  fprintf (stderr, "Progress %ld / %ld\n", a, b);
  return 0;
}


static error_t parse_opt (int key, char *arg, struct argp_state *state)
{
  (void) arg;
  (void) state;
  switch (key)
    {
    case 'v': verbose++;
      debuginfod_set_progressfn (client, & progressfn); break;
    default: return ARGP_ERR_UNKNOWN;
    }
  return 0;
}


/* Data structure to communicate with argp functions.  */
static struct argp argp =
  {
   options, parse_opt, args_doc, doc, NULL, NULL, NULL
  };



int
main(int argc, char** argv)
{
  elf_version (EV_CURRENT);

  client = debuginfod_begin ();
  if (client == NULL)
    {
      fprintf(stderr, "Couldn't create debuginfod client context\n");
      return 1;
    }

  /* Exercise user data pointer, to support testing only. */
  debuginfod_set_user_data (client, (void *)"Progress");

  int remaining;
  (void) argp_parse (&argp, argc, argv, ARGP_IN_ORDER|ARGP_NO_ARGS, &remaining, NULL);

  if (argc < 2 || remaining+1 == argc) /* no arguments or at least two non-option words */
    {
      argp_help (&argp, stderr, ARGP_HELP_USAGE, argv[0]);
      return 1;
    }

  int rc;
  char *cache_name;

  /* Check whether FILETYPE is valid and call the appropriate
     debuginfod_find_* function. If FILETYPE is "source"
     then ensure a FILENAME was also supplied as an argument.  */
  if (strcmp(argv[remaining], "debuginfo") == 0)
    rc = debuginfod_find_debuginfo(client,
				   (unsigned char *)argv[remaining+1], 0,
				   &cache_name);
  else if (strcmp(argv[remaining], "executable") == 0)
    rc = debuginfod_find_executable(client,
				    (unsigned char *)argv[remaining+1], 0,
				    &cache_name);
  else if (strcmp(argv[remaining], "source") == 0)
    {
      if (remaining+2 == argc || argv[remaining+2][0] != '/')
        {
          fprintf(stderr, "If FILETYPE is \"source\" then absolute /FILENAME must be given\n");
          return 1;
        }
      rc = debuginfod_find_source(client, (unsigned char *)argv[remaining+1],
				  0, argv[remaining+2], &cache_name);
    }
  else
    {
      argp_help (&argp, stderr, ARGP_HELP_USAGE, argv[0]);
      return 1;
    }

  if (verbose)
    {
      const char* url = debuginfod_get_url (client);
      if (url != NULL)
        fprintf(stderr, "Downloaded from %s\n", url);
    }

  debuginfod_end (client);

  if (rc < 0)
    {
      fprintf(stderr, "Server query failed: %s\n", strerror(-rc));
      return 1;
    }

  printf("%s\n", cache_name);
  free (cache_name);

  return 0;
}
